### PR TITLE
feat(MarkdownContent): Expose `transformLinkUri` and `transformImageUri`

### PR DESCRIPTION
**Pull request by Rowern**
## Hey, I just made a Pull Request!

This pull request exposes the `transformLinkUri` and `transformImageUri` props of react-markdown, this can be exceptionally useful to re-write URLs when using the `MarkdownContent` for internal content as well as images (ie changing relative imports to other URLs like a CDN or something else).

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by jhaals — COMMENTED** [review]
Thanks for the PR alongside the context 👍 
Would it be possible to define our own types for `transformImageUri` and `transformLinkUri`? My hope is that those can be fairly small and that we avoid exporting types from external dependencies in our API :)

**Comment by jhaals** [comment]
> > My hope is that those can be fairly small and that we avoid exporting types from external dependencies in our API :)
> 
> You're right, I moved to a "simpler" type but this means that there will be an `any[]` because the actual type is too complex to replicate it in backstage. What do you think ?

Would it be possible to only include the things needed from the type like href and source and skip the rest?

**Comment by Rowern** [comment]
> Would it be possible to only include the things needed from the type like href and source and skip the rest?

I'm not really sure how I could simplify this type:
```ts
type TransformLink = (
  href: string,
  // Complex type from internal react-markdown dep library (hast)[./node_modules/@types/hast/index.d.ts]
  children: any[],
  title: string | null,
) => string;
```
I don't really want to remove the second argument because eventhough the typings is not optimal some user might want to use it in their transformations.
Would this solution be acceptable:

```ts
type TransformLink = (
  href: string,
  // Complex type from react-markdown
  _: any,
  title: string | null,
) => string;
```

**Comment by Rugvip** [comment]
Let's simplify it to this for now:

```ts
type TransformLink = (href: string) => string;
```

If it turns out that someone really needs the other arguments we can add them back later, but to me they seem to very much to cause a tricky coupling between our API surface and the dependency. Right now a major bump to the hast tree really means we need to do a major bump of `core-components`, which is far from ideal.

**Comment by Rowern** [comment]
@Rugvip Should be good now :)

File: packages/core-components/src/components/MarkdownContent/MarkdownContent.tsx [hunk 1]
```diff
@@ -64,10 +64,18 @@ const useStyles = makeStyles<BackstageTheme>(
...
 
+type TransformLink = (href: string) => string;
+
 type Props = {
   content: string;
   dialect?: 'gfm' | 'common-mark';
-  linkTarget?: Options['linkTarget'];
+  linkTarget?: React.HTMLAttributeAnchorTarget | TransformLink;
+  transformLinkUri?: TransformLink;
+  transformImageUri?: (
```
Comment (Rugvip) [inline packages/core-components/src/components/MarkdownContent/MarkdownContent.tsx:74]: Let's have both of these use just `(href: string) => string` inlined, not feeling it's worth the extra named type tbh, as we'd need to export that.

**Review by benjdlambert — APPROVED** [review]
Looking good! Thanks for this! 👍

**Pull request merged**